spec.bs

```diff
@@ -192,31 +327,59 @@ Specific examples:
-To change the [=Document/frozenness=] of a top-level document passing in a document |doc| and [=Document/frozenness=] state |x| (one of <code>true</code>, <code>false</code>):
-
-1. Assert: |doc|'s [=Document/browsing context=] is a [=top-level browsing context=].
-1. Execute [[#change-frozenness-of-document]] passing in |doc| and |x|.
-1. Let |descendants| be the [=list of the descendant browsing contexts=] of |doc|.
-1. For each [=/browsing context=] |b| in |descendants|:
-    1. Let |d| be the [=active document=] of |b|.
-    1. Execute [[#change-frozenness-of-document]] passing in |d| and |x|.
-
-<h5 dfn>Change frozenness of document</h5>
-To change the [=Document/frozenness=] of a document given a document |doc| and [=Document/frozenness=] state |x|:
-1. If |x| is <code>true</code>:
-    1. Set |doc|'s [=Document/frozenness=] state to <code>true</code>.
+The UA will typically execute [=change the frozenness of a top-level document=] with false when the user revisits that browsing context. In addition, the UA may choose to periodically execute [=change the frozenness of a top-level document=] with false in the background, if plentiful resources are available.
+
+<h4 id="changing-frozenness">Changing the frozenness of documents</h4>
+
+<div algorithm>
+    To <dfn>change the frozenness of a top-level document</dfn>, given a {{Document}} |topLevelDoc| and boolean [=Document/frozenness=] state |frozenness|:
+
+    1. Assert: |doc|'s [=Document/browsing context=] is a [=top-level browsing context=].
+    1. [=Change the frozenness of a document=] given |topLevelDoc|, |frozenness|, and false.
+    1. Let |descendants| be the [=list of the descendant browsing contexts=] of |doc|.
+    1. For each [=/browsing context=] |b| in |descendants|:
+        1. Let |descendantDocument| be the [=active document=] of |b|.
+        1. [=Change the frozenness of a document=] given |descendantDocument|, |frozenness|, and false.
+</div>
+
+<div algorithm>
+    To <dfn>change the frozenness of a document</dfn>, given a {{Document}} |doc|, a boolean [=Document/frozenness=] state |frozenness|, and a boolean |auto resume frozen media|:
+
+    1. If |frozenness| is true, run the [=freeze steps=] for |doc| given |auto resume frozen media|.
+    1. Otherwise, run the [=resume steps=] given |doc|.
+</div>
+
+<div algorithm>
+    To run the <dfn>freeze steps</dfn> for a {{Document}} <var>doc</var>, given a boolean |auto resume frozen media|:
+
+    1. Set |doc|'s [=Document/frozenness=] state to true.
     1. [=Fire an event=] named <code>freeze</code> at |doc|.
-1. Otherwise:
+    1. Let |elements| be all [=media elements=] in |doc|, in [=tree order=].
+    1. For each |element| in |elements|:
+        1. If |element|'s {{HTMLMediaElement/paused}} is false, then:
+            1. Set |element|'s [=HTMLMediaElement/resume frozen flag=] to |auto resume frozen media|.
+            1. Call |element|'s {{HTMLMediaElement/pause()}} method.
```

Comment:
Calling public methods is not good. This will invoke arbitrary code if someone has done `HTMLMediaElement.prototype.pause = function () { ... }`.

Instead you should invoke the underlying algorithm. Ideally you would invoke the internal pause steps: https://html.spec.whatwg.org/multipage/media.html#internal-pause-steps

But you linked to something that also performs special behavior for NETWORK_EMPTY, doing extra work. Do you want to do that extra work, or do the internal pause steps suffice?

Reply:
Ya we excplicitly call the C++ code for this but I didn't know how to write that in the spec. But ya calling the steps itself is better..

Reply:
The C++ code invokes these code paths. I have mapped the play, pause directly to the top level algorithms and not the IDL binding.

Reply:
Done